src/System.Linq/tests/ZipTests.cs

```diff
@@ -0,0 +1,393 @@
+using Xunit;
+
+namespace System.Linq.Tests
+{
+    public class ZipTests
+    {
+        [Fact]
+        public void ImplicitTypeParameters()
+        {
+            IEnumerable<int> first = new int[] { 1, 2, 3 };
+            IEnumerable<int> second = new int[] { 2, 5, 9 };
+            IEnumerable<int> expected = new int[] { 3, 7, 12 };
+
+            Assert.Equal(expected, first.Zip(second, (x, y) => x + y));
+        }
+        
+        [Fact]
+        public void ExplicitTypeParameters()
+        {
+            IEnumerable<int> first = new int[] { 1, 2, 3 };
+            IEnumerable<int> second = new int[] { 2, 5, 9 };
+            IEnumerable<int> expected = new int[] { 3, 7, 12 };
+
+            Assert.Equal(expected, first.Zip<int, int, int>(second, (x, y) => x + y));
+        }
+
+        [Fact]
+        public void FirstIsNull()
+        {
+            IEnumerable<int> first = null;
+            IEnumerable<int> second = new int[] { 2, 5, 9 };
+            
+            Assert.Throws<ArgumentNullException>(() => first.Zip<int, int, int>(second, (x, y) => x + y));
+        }
+
+        [Fact]
+        public void SecondIsNull()
+        {
+            IEnumerable<int> first = new int[] { 1, 2, 3 };
+            IEnumerable<int> second = null;
+            
+            Assert.Throws<ArgumentNullException>(() => first.Zip<int, int, int>(second, (x, y) => x + y));
+        }
+        
+        [Fact]
+        public void FuncIsNull()
+        {
+            IEnumerable<int> first = new int[] { 1, 2, 3 };
+            IEnumerable<int> second = new int[] { 2, 4, 6 };
+            Func<int, int, int> func = null;
+            
+            Assert.Throws<ArgumentNullException>(() => first.Zip(second, func));
+        }
+
+        private class MyIEnum<T> : IEnumerable<T>
+        {
+            public IEnumerable<T> _data;
+
+            public MyIEnum(IEnumerable<T> source)
+            {
+                _data = source;
+            }
+
+            public IEnumerator<T> GetEnumerator()
+            {
+                foreach (var datum in _data)
+                {
+                    if (datum.Equals(2)) throw new Exception();
+                    yield return datum;
+                }
+            }
+
+            IEnumerator IEnumerable.GetEnumerator()
+            {
+                return this.GetEnumerator();
+            }
+        }
+
+        [Fact]
+        public void ExceptionThrownFromFirstGetEnumerator()
+        {
+            MyIEnum<int> first = new MyIEnum<int>(new int[] { 1, 3, 3 });
+            IEnumerable<int> second = new int[] { 2, 4, 6 };
+            Func<int, int, int> func = (x, y) => x + y;
+            IEnumerable<int> expected = new int[] { 3, 7, 9 };
+            
+            Assert.Equal(expected, first.Zip(second, func));
+
+            first = new MyIEnum<int>(new int[] { 1, 2, 3 });
+            
+            Assert.ThrowsAny<Exception>(() => first.Zip(second, func).ToList());
```

Comment:
Should the first.Zip move to before the ThrowsAny to make it clear that the exception is coming from the enumeration and not from the operator itself?